bundles/org.eclipse.e4.ui.css.swt/src/org/eclipse/e4/ui/css/swt/helpers/CSSSWTColorHelper.java

```diff
@@ -353,8 +354,20 @@ private static RGBA findColorByDefinition(String name) {
 	/** Helper function to avoid setting colors unnecessarily */
 	public static void setForeground(Control control, Color newColor) {
```

Comment:
up to now this method only skipp the setting if the old and the new color where identical.
With your addition this method now does a lot more. Can you explain why this "isCLabelInFormHeading" is needed. It feels a bit wrong to have control specific logic inside this class.

Reply:
The problem is that setting the color to <unset> in the CSS file will result in newColor=null. In this case, the control.setForegorund would use the default system color. But for CLabel in form headings we want to keep the color that is set by the message.
The only way (at least that I can see) to achieve this is to avoid that the color is set if newColor==null && control is a CLabel in a form heading. In all other cases we don't want to change the behavior.

Reply:
> we want to keep the color that is set by the message.

How is the color set by the message? In code or via CSS?

Reply:
It's set via coding in FormHeading based on the messageType:
private void updateForeground() {
			Color theFg;

			switch (messageType) {
			case IMessageProvider.ERROR:
				theFg = getDisplay().getSystemColor(SWT.COLOR_RED);
				break;
			case IMessageProvider.WARNING:
				theFg = getDisplay().getSystemColor(SWT.COLOR_DARK_YELLOW);
				break;
			default:
				theFg = fg;
			}
			getMessageControl().setForeground(theFg);
		}

Reply:
ah ok. Does SWT.COLOR_RED and SWT.COLOR_DARY_YELLOW change their "effective" color "under the hood" via CSS?

Reply:
No, the colors are hard coded and are the same for every theme

Reply:
Mh. Does is look good in both the dark and the light theme?

Reply:
It's not perfect but I think It's reasonably okay
<img width="646" alt="MessageColors" src="https://github.com/eclipse-platform/eclipse.platform.ui/assets/79514265/810774bb-23af-45ae-9000-38ea61a6317f">

Reply:
Mh. The red is really bright in the dark version and the black has a bad contrast in the dark version. I am wondering if your initial issue would also be fixed by defining new 3 new (semantic) colors (in the colors and fonts preference dialog). Giving the good values for the dark theme and instead of setting the system colors in code setting the new semnatic colors.

Reply:
This would not solve the initial issue. The problem is that the CSSEngine set's the color of the CLabel in the Form Heading to a specified hard coded color, the CSSEngine has no knowledge about the message kind. Therefore changing the color via the theme is not possible. 

Of course we can add new colors in the preferences to change the colors when the message is added, but as I said this will not fix the issue in the CSSEngine.

Reply:
Wouldn't it be better to define semantic colors for error, warning in info. Define meaningful colors for the light and dark theme and setting these semantic colors instead of setting the system colors?

You could add the "messageType" as an CSS class to the Control e.g. something like this:

```
// set the class, can be used several times
label.setData("org.eclipse.e4.ui.css.CssClassName", "warning")
```

In the CSS you could then define something like "a CLabel with class 'error' should have color red" and "a CLabel with class "warning" should have color yellow" in the CSS
In addition to setting the color in the updateForeground() method you would set the class on the CLabel.

See also https://www.vogella.com/tutorials/Eclipse4CSS/article.html#styling-based-on-css-class-or-css-id for more details

The "color" property of the .MPart FormHeading > CLabel CSS-Rule (which you already adapted in your PR) could then be removed completely. This rule is wrong anyways (because messages in the FormHeading are not always error messages).

Reply:
Thanks four your help, the suggestion looks very promising. Here are the new color settings in dark theme. I'm only struggling with the info color, but did not find a better option.
<img width="510" alt="MessageColors" src="https://github.com/eclipse-platform/eclipse.platform.ui/assets/79514265/e4d05480-f25d-40e9-84dc-19b79c0f763d">

Reply:
> only struggling with the info color

What do you mean with "struggling"? Finding a "better" color value?

Reply:
Maybe the info color should be more light - more similar to the color used in the editor tab.

Reply:
Changed it to the editor color, looks much better now: 
<img width="465" alt="Bildschirmfoto 2023-08-07 um 11 40 36" src="https://github.com/eclipse-platform/eclipse.platform.ui/assets/79514265/532b5a28-d22f-42ed-95b5-cd375004566b">